README.md
### Golf Conditions Application

This application uses the [aeris weather api](https://www.aerisweather.com/support/docs/api/reference/endpoints) to fetch weather conditions at a given location and calculates a rating based on a few factors of how ideal the weather is for playing golf.

Things analyzed are:
 * wind (30% weight)
 * temperature (30% weight)
 * precipitation (25% weight)
 * dew point (15% weight)


## The Backend API

The backend for this app was built using Python's [FastAPI](https://fastapi.tiangolo.com/) microframework.  There are two basic services, one for Weather Conditions and another to serve up Golf Courses in Minnesota.

Before you can run this project, you must make sure you have an account with [aeris weather](https://www.aerisweather.com/signup/developer/) and [mapbox](https://docs.mapbox.com/help/getting-started/access-tokens/) and you will need to create a couple `.env` files.  Create this file in and fill in the necessary variables at `/app/routers/weather/.env`:

```env
# /app/routers/weather/.env
ARIES_CLIENT_ID=<your-id>
ARIES_CLIENT_SECRET=<your-secret>
```

To run the backend:

```sh
# source your profile
source ~/.bash_profile
# activate a virtual environment
conda activate <some virtual environment>
# install the requirements for this project using the included requirements.txt file
pip install -r requirements.txt
# run app with uvicorn
uvicorn app.main:app --reload
```

By default, this will spin up on port `8000` at [http://localhost:8000](http://localhost:8000/) or you may have to use the local address (`127.0.0.1:8000`).  To view the [Swagger](https://swagger.io/tools/swagger-ui/) docs, visit:

[http://127.0.0.1:8000/docs](http://127.0.0.1:8000/docs)

![swagger](./resources/images/swagger.png)

or if you prefer [ReDoc](https://redocly.github.io/redoc/):

[http://127.0.0.1:8000/redoc](http://127.0.0.1:8000/redoc)

![redoc](./resources/images/redoc.png)

#### golf course data

The golf course data is stored in a json file [here](./app/data/All_MN_Courses.json). This data was derived from manually downloading a few json payloads from [MinnesotaGolf.com](https://minnesotagolf.com/fairways), and then using a [script](./app/scripts/crawler.js) to scrape and join data for each region and finally merge all the regions together into one data source.

This does support a quick and dirty dictionary based query via url params. For example, to find golf courses within Minneapolis:

[http://127.0.0.1:8000/golf-courses?City=Minneapolis](http://127.0.0.1:8000/golf-courses?City=Minneapolis)


### front end app

There is a [front end app](./golf-conditions/) available to use the API to find golf conditions at any given location.  When clicking on the map, a request is made to the [aeris weather conditions endpoint](https://www.aerisweather.com/support/docs/api/reference/endpoints/conditions/) looking at the next 6 hours to calculate a suitability ranking on a 1-5 scale (1=bad, 5=good) to determine the conditions for playing golf at that location:

![golf index request](./resources/images/golf-index-request.png)


When the [service](./app/main.py) is running, the app should be served at the default location of [http://127.0.0.1:8000/app](http://127.0.0.1:8000/app) (may have to use `localhost:8000/app`) where you can click on a golf course to see the current condition rating:

![front end app](./resources/images/front-end-app.png)

You can also use the `Geocoder` in the upper right hand corner to check on areas that are nice for playing golf year around:


![san diego](./resources/images/san-diego.png)

#### front end app setup

To modify the front end app source code, some additional setup is requried. Make the following env file at `/golf-conditions/.env`:

```
# /golf-conditions/.env
VITE_MAPBOX_TOKEN=<your-token>
VITE_AERIS_CLIENT_ID=<your-client>
VITE_AERIS_CLIENT_SECRET=<your-secret>
VITE_APP_PUBLIC_PATH=.
```

and then a production one, where you can change the public path if hosting from an actual server:

```
# /golf-conditions/.env.production
VITE_APP_PUBLIC_PATH=..
```

then cd into the `/golf-conditions` directory and install dependencies:

```sh
cd golf-conditions
npm i
npm run dev
```

In order to build the `dist` files for the backend services, you will also need to run:

```sh
npm run build
```

This will automatically deploy the transpiled code to the `app/static` folder and can then be served through the `/app` endpoint.

#### about the front end app

This app was build with Vue.js, TypeScript, Vite and [vue-mapbox-ts](https://gitlab.com/relief-melone/vue-mapbox-ts).

#### backend tests

test the ratings:

```sh
python -m unittest app/test/test_aeris.py
```

test the conditions (mock):
```sh
python -m pytest app/test/test_mocks.py
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.2.1](https://github.com/CalebM1987/golf-conditions/compare/v1.2.0...v1.2.1) (2022-10-22)

## [1.2.0](https://github.com/CalebM1987/golf-conditions/compare/v1.1.0...v1.2.0) (2022-10-22)


### Features

* add limit and use dropdown for format ([de25044](https://github.com/CalebM1987/golf-conditions/commit/de250442cdb86a859df305f9b2ec74182c455f03))
* **weather-info:** show wind speed when >= 10kph ([7ee9f33](https://github.com/CalebM1987/golf-conditions/commit/7ee9f336b1e1a72bfc997f95e3751a9c7b4f4b20))

## [1.1.0](https://github.com/CalebM1987/golf-conditions/compare/v1.0.1-0...v1.1.0) (2022-10-22)


### Features

* **golf-courses:** make all requests async ([ea31b44](https://github.com/CalebM1987/golf-conditions/commit/ea31b448786c958e6e4c5a05fedf79c6bac4b02d))


### Bug Fixes

* **aeris:** fixed get_args and improved rating logic ([077a302](https://github.com/CalebM1987/golf-conditions/commit/077a30275e202ece07ab2f2335b2bc1046de6fe8))

### 1.0.1-0 (2022-10-20)


### Features

* **app:** serve front end from /app endpoint ([4b1e5bd](https://github.com/CalebM1987/golf-conditions/commit/4b1e5bdca095d75c6eadb2d4a1b44a38c2cace03))
* set up initial state ([59de891](https://github.com/CalebM1987/golf-conditions/commit/59de8914086fb865ea601c3283e0de5737a2b0a6))
* show selected golf course ([6dfd406](https://github.com/CalebM1987/golf-conditions/commit/6dfd4060badbac7efde735bdb79d11903fd651dd))
* show weather info and activity rating ([fe8d53e](https://github.com/CalebM1987/golf-conditions/commit/fe8d53e648d207bbdb208262e1e73f4569afdf6e))
* update weather on geocode result ([6504e3f](https://github.com/CalebM1987/golf-conditions/commit/6504e3f15954b220110397d40f96ea109f21c46f))
* **weather:** added typings for golf ratings ([2227d42](https://github.com/CalebM1987/golf-conditions/commit/2227d4294dc36e3d29d63eed2b59962fa080059c))


### Bug Fixes

* **weather:** match other indices schemas ([163dbc0](https://github.com/CalebM1987/golf-conditions/commit/163dbc08069ad900f4240b0126ba4c63708437d4))
* **weather:** return valid response on error ([ccaa5a3](https://github.com/CalebM1987/golf-conditions/commit/ccaa5a33c25e0a5b1bd02ecc64273b95514aa5b3))


app/routers/weather/aeris.py
import os
import httpx
import math
from typing import Union, List
from munch import munchify
from dotenv import dotenv_values
from .models import RatingResponse, WeatherConditionsResponse
from .ratings import *
from app.utils import get_timestamp, from_milliseconds

dotfile = os.path.join(os.path.abspath(os.path.dirname(__file__)), '.env')
aries_env = dotenv_values(dotfile)

# aries base url
base = 'https://api.aerisapi.com'

# rating helpers
RATING_CATEGORIES = ['bad', 'poor', 'good', 'very good', 'excellent']
RATINGS = dict((i+1, c) for i,c in enumerate(RATING_CATEGORIES))
RATINGS_REVERSED = dict((i+1, c) for i,c in enumerate(RATING_CATEGORIES[::-1]))

# filters for weather fields
weather_fields =  [
    'loc', 
    'place', 
    'periods.timestamp',
    'periods.dateTimeISO',
    'periods.range', 
    'periods.precip', 
    'periods.temp', 
    'periods.dewpoint',
    'periods.windSpeed',
    'periods.weather.phrase',
    'periods.weather.primary',
    'profile.tz'
]

def get_args(f='json', fields: Union[str, List[str]]=None, **kwargs):
    """combine aeris client auth params with other query parameters

    Args:
        f (str, optional): the request format. Defaults to 'json'.
        fields (Union[str, List[str]], optional): list of fields to filter. Defaults to None.

    Returns:
        _type_: _description_
    """
    CLIENT_ID = aries_env.get('AERIS_CLIENT_ID')
    CLIENT_SECRET = aries_env.get('AERIS_CLIENT_SECRET')
    params = dict(
        f=f,
        client_id=CLIENT_ID,
        client_secret=CLIENT_SECRET,
        **kwargs
    )

    # validate fields
    if fields:
        if isinstance(fields, list):
            fields = ','.join(fields)
        params['fields'] = fields 
   
    return params


async def get_current_conditions(location: str, fields=weather_fields, **kwargs) -> WeatherConditionsResponse:
    """query the aeris weather api to fetch conditions for a given location

    Args:
        location (str): _description_

    Returns:
        _type_: _description_
    """
    url = f'{base}/conditions/summary/{location}'
    
    async with httpx.AsyncClient() as client:
        r = await client.get(url, params=get_args(fields=fields, **kwargs))
        return munchify(r.json())


async def get_golf_conditions(location: str, reverse=False) -> RatingResponse:
    """calculates a golf condition ranking from 1-5, based on weighted criteria

    https://theweatherprediction.com/habyhints2/606/

    # weights
    # 1. temp close to 72 degrees F, (25%)
    # 2. wind under 12 mph, lower the better (%30)
    # 3. precipitation (light: <.5mm/hr, mod: > .5 & < 4mm/hr, heavy, >4mm/hr) (30%)
    # 4. dew point, under 50 is comfortable lower the better (15%)

    # rank is round(weightedScore / 20)

    If temp is less than 40, the rating will automatically be set to 1

    Args:
        location (str): a place or lat/lng

    Returns:
        int: the golf condition ranking
    """
    
    error = None
    rating = None
    _min, _max = 1, 5
    indexNames = RATINGS
    period = None
    conditions = {}
    keep = ['loc', 'place', 'profile']
    if reverse:
        _min, _max = 5, 1
        indexNames = RATINGS_REVERSED
    try:
        resp = await get_current_conditions(location, to='+6hours')
        conditions = resp.response[0]
        period = conditions.periods[0]

        # get condition elements
        tempF = period.temp.avgF
        if tempF < 40:
            rating = 1
        
        else:
            windMph = period.windSpeed.avgMPH
            precipMM = period.precip.avgMM
            dewF = period.dewpoint.avgF

            # get temp ratings
            temp = get_temp_rating(tempF) * 25
            wind = get_wind_rating(windMph, tempF) * 30
            precip = get_precip_rating(precipMM) * 30
            dew = get_dew_rating(dewF, tempF) * 15

            # get rank
            rating = min(5, round(sum([temp, wind, precip, dew]) / 20)) 
        if reverse:
            rating = 6 - (rating or 1)
            
    except Exception as e:
        error = str(e)
        ts = get_timestamp()
        period = munchify(dict(
            timestamp=ts,
            dateTimeISO=from_milliseconds(ts)
        ))
    
    # return rating and other info
    return munchify(dict(
        success=error == None,
        error=error,
        response=[
            dict(
                indice=dict(
                    type="golf",
                    range=dict(
                        min=_min,
                        max=_max,
                        reverse=reverse
                    ),
                    past=None,
                    current=dict(
                        index=rating,
                        timestamp=period.timestamp,
                        dateTimeISO=period.dateTimeISO,
                        indexENG=indexNames.get(rating)
                    ),
                ),
                forecast=None, # not sure what this is?
                **{k:v for k,v in conditions.items() if k in keep}
            ),
        ]
    ))
